### PR TITLE
[go/conformance] Fix l2-resource-primitives test

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
### Description

This PR fixes the l2-resource-primitives conformance test in Go. The problem was that expressions of the form `UnaryOperationExpr(Negate, NumberExpr(N))` didn't get _lifted_ properly: we emitted `-N` instead of `pulumi.Float64(-N)`. 

The solution here is to rewrite the expression first from `UnaryOperationExpr(Negate, NumberExpr(N))`  into `NumberExpr(-N)` and only then lift it like any other number.

### CODE REVIEW

File: pkg/codegen/go/gen_program_expressions.go [hunk 1]
```diff
@@ -1022,12 +1022,44 @@ func (nameInfo) Format(name string) string {
 	return name
 }
 
+// rewriteUnaryNegate rewrites UnaryOperation(-, Number(N)) into Number(-N).
+func rewriteUnaryNegate(expr model.Expression) (model.Expression, hcl.Diagnostics) {
```
Comment (Frassle) [inline pkg/codegen/go/gen_program_expressions.go:1026]: I think we should just lift this to pcl binding, not make it Go specific. I had to do the same thing for the other languages when adding bigint support and just did this simplification in bindUnaryOpExpression (https://github.com/pulumi/pulumi/pull/16939/files#diff-0925bf1716dbbe0ef4c7363db8d274d28213f80d7620a1a6a6213cd775c94bf5R786).

Feels like we may as well do that now, as it fixes this for Go and will be needed for bigints later anyway.

Comment (Zaid-Ajaj) [reply pkg/codegen/go/gen_program_expressions.go]: Done ✅

File: pkg/codegen/hcl2/model/binder_expression.go [hunk 2]
```diff
@@ -769,6 +769,23 @@ func (b *expressionBinder) bindUnaryOpExpression(syntax *hclsyntax.UnaryOpExpr)
 	if tokens == nil {
 		tokens = _syntax.NewUnaryOpTokens(syntax.Op)
 	}
+
+	// simplify UnaryOperation(Negate, Number(N)) into Number(-N)
+	if syntax.Op == hclsyntax.OpNegate {
+		if lit, ok := operand.(*LiteralValueExpression); ok && lit.Value.Type().Equals(cty.Number) {
+			num, _ := lit.Value.AsBigFloat().Int64()
+			negated := &LiteralValueExpression{
+				Value:  cty.NumberIntVal(-num),
```
Comment (Frassle) [inline pkg/codegen/hcl2/model/binder_expression.go:778]: Do we even need to go via int64 for this? Can't we stay as bigFloat given that's what cty is gonna work with internal anyway. Also means this is correct for things like -122.30, which we should probably ensure we have a test for!

Comment (Zaid-Ajaj) [reply pkg/codegen/hcl2/model/binder_expression.go:778]: I think we might be able to just `.Negate()` 🤔 will double check

File: pkg/codegen/hcl2/model/binder_expression.go [hunk 3]
```diff
@@ -769,6 +769,22 @@ func (b *expressionBinder) bindUnaryOpExpression(syntax *hclsyntax.UnaryOpExpr)
 	if tokens == nil {
 		tokens = _syntax.NewUnaryOpTokens(syntax.Op)
 	}
+
+	// simplify UnaryOperation(Negate, Number(N)) into Number(-N)
+	if syntax.Op == hclsyntax.OpNegate {
+		if lit, ok := operand.(*LiteralValueExpression); ok && lit.Value.Type().Equals(cty.Number) {
+			negated := &LiteralValueExpression{
+				Value:  lit.Value.Negate(),
+				Syntax: lit.Syntax,
```
Comment (Frassle) [inline pkg/codegen/hcl2/model/binder_expression.go:778]: Your losing the syntax and tokens of the operator here, and Syntax.Val will be the wrong value. Again I think you can just copy what I did in https://github.com/pulumi/pulumi/pull/16939/files#diff-0925bf1716dbbe0ef4c7363db8d274d28213f80d7620a1a6a6213cd775c94bf5R786

Comment (Zaid-Ajaj) [reply pkg/codegen/hcl2/model/binder_expression.go:778]: > Your losing the syntax and tokens of the operator here

Nice catch!

Comment (Zaid-Ajaj) [reply pkg/codegen/hcl2/model/binder_expression.go]: Done ✅

**Review by Frassle — APPROVED** [review]

**Pull request merged**